contracts/metatx/MinimalForwarder.sol

```diff
@@ -23,82 +29,202 @@ contract MinimalForwarder is EIP712 {
+     * will receive the requested gas and no ETH is stuck in the contract.
+     */
     function execute(
-        ForwardRequest calldata req,
+        ForwardRequest calldata request,
         bytes calldata signature
-    ) public payable returns (bool, bytes memory) {
-        address signer = _recover(req, signature);
-        (bool correctFrom, bool correctNonce) = _validateReq(req, signer);
+    ) public payable virtual returns (bool, bytes memory) {
+        (bool success, bytes memory returndata) = _execute(request, signature);
 
-        if (!correctFrom) {
-            revert MinimalForwarderInvalidSigner(signer, req.from);
+        if (msg.value != request.value) {
+            revert MinimalForwarderMismatchedValue(request.value, msg.value);
         }
-        if (!correctNonce) {
-            revert MinimalForwarderInvalidNonce(signer, _nonces[req.from]);
+
+        return (success, returndata);
+    }
+
+    /**
+     * @dev Batch version of {execute}.
+     */
+    function executeBatch(ForwardRequest[] calldata requests, bytes[] calldata signatures) public payable virtual {
+        if (requests.length != signatures.length) {
+            revert MinimalForwarderInvalidBatchLength(requests.length, signatures.length);
         }
 
-        _nonces[req.from] = req.nonce + 1;
+        uint256 requestsValue;
 
-        (bool success, bytes memory returndata) = req.to.call{gas: req.gas, value: req.value}(
-            abi.encodePacked(req.data, req.from)
-        );
+        for (uint256 i; i < requests.length; ++i) {
+            requestsValue += requests[i].value;
+            _execute(requests[i], signatures[i]);
+        }
 
-        // Validate that the relayer has sent enough gas for the call.
-        // See https://ronan.eth.limo/blog/ethereum-gas-dangers/
-        if (gasleft() <= req.gas / 63) {
-            // We explicitly trigger invalid opcode to consume all gas and bubble-up the effects, since
-            // neither revert or assert consume all gas since Solidity 0.8.0
-            // https://docs.soliditylang.org/en/v0.8.0/control-structures.html#panic-via-assert-and-error-via-require
-            /// @solidity memory-safe-assembly
-            assembly {
-                invalid()
-            }
+        if (msg.value != requestsValue) {
+            revert MinimalForwarderMismatchedValue(requestsValue, msg.value);
         }
+    }
 
-        return (success, returndata);
+    /**
+     * @dev Validates if the provided request can be executed at current block with `signature` on behalf of `signer`.
+     */
+    function _validate(
+        ForwardRequest calldata request,
+        bytes calldata signature
+    ) internal view virtual returns (bool alive, bool signerMatch, bool nonceMatch) {
+        address signer = _recoverForwardRequestSigner(request, signature);
+        return (request.deadline >= block.number, signer == request.from, nonces(request.from) == request.nonce);
```

Comment:
Do we have a rational for using `block.number` for the deadline ? Historically we always did that, but we also now today that they are not really predicable, particularly on L2s. Should we use block.timestamp instead ?

Reply:
I like the idea but I think it makes sense to block because transactions are generally block-number sensitive (eg. Flashbots). I don't see a major reason to keep `block.number` so I'd go for `block.timestamp` since ERC20Permit uses timestamp as well.

Shall we consider a `clock()`? CC @frangio

Reply:
I'm not sure how flashbots affect this. My understanding is that if you put a deadline its because you don't the message to sit around for weeks, and execute when the context is not favorable (for example a buy order at a price that is ok now, but that may be very low in 3 months).

When you want your messaged to be relayed, flashbot can happen before or after ... just like a normal transaction. I don't see how they are relevant here. Even if they are, how do you know in which block number a flashbot will operate ?

I may not see the full picture, but IMO the deadline is mostly a user thing: "if that is not done by sunday, then forget it". In that regard, timestamps are way better IMO.

Not sure about the clock. It would work, but it would also require the UI that prepare the request to do clock detection.

Reply:
Agree on using timestamps because they have better usability. Using block number does not mitigate MEV as far as I can tell (it's still possible to frontrun and reorder within one block).

I don't see a need for a customizable clock either, and I agree it would make it harder for tooling since they would have to detect what clock is being used.

Reply:
> I'm not sure how flashbots affect this.

It doesn't, what I say is that transactions might be targeted to a block. For example, a request signer might want to limit its request to a certain block because there's an MEV concern in further blocks. For such cases, a timestamp doesn't offer the same guarantees.

Although I generally agree it's an edge case and timestamps are fine.

I'll update with the timestamp